### PR TITLE
Create README.md

### PR DESCRIPTION
**Pull request by TashaSekularac**

### CODE REVIEW

File: assignment2.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+//
+
+#include <iostream>
+#include <stack>
+
+using namespace std;
+
+struct Node
+{
+public:
```
Comment (Brusi) [inline assignment2.cpp:16]: Struct fields are public by default. This is actually the only difference between `class` and `struct` keywords.
You can remove the `public:`.

File: assignment2.cpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+void printAllAncestors(Node *root, Node* target)
+{
+    printAllAncestorsHelp(root, target);
+    cout<<endl;
+}
+
+
+
+
+bool cover(Node* root, Node* target)
```
Comment (Brusi) [inline assignment2.cpp:48]: This function name is not clear.
How about `contains` ? (as if the `root` is the root of a tree that contains the `target` node).
Or `isAncestorOf`? (returns whether `root` is an ancestor of `target`)?

File: assignment2.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    if ( printAllAncestorsHelp(root->left, target) || printAllAncestorsHelp(root->right, target) )
+    {
+        cout << root->val << " ";
+        return true;
+    }
+    return false;
+}
+
+
+void printAllAncestors(Node *root, Node* target)
```
Comment (Brusi) [inline assignment2.cpp:39]: Assignment 1 begins with the words "Given a Binary Tree and a key", meaning that it is preferred that the input for this function is a node and a key (int).
Luckily, looking at your solution, you never really compared the 'target' node by its identity, but only compared its `val` field. So changing this to work with a key (instead of a node) would not require big changes and will result in clearer code :)

File: assignment2.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    
+    if (x != nullptr && y!=nullptr)
+            return root;
+    if (root->val ==p->val || root->val == q->val)
+        return root;
+    if(x==nullptr)
+        return y;
+    return x;
+}
+Node* commonAncestor(Node* root, Node* p, Node* q)
```
Comment (Brusi) [inline assignment2.cpp:77]: Though the assignment did not state that explicitly, I think that it is better that this function will get a node and two keys as its input (instead of one 'root' node and two nodes to search their ancestors).
Again here, I think the code will look better with it :)

File: assignment2.cpp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    if (root==nullptr) return nullptr;
+    
+    Node* x = commonAncestorHelp(root->left, p, q);
+    if (x != nullptr && x->val != p->val && x ->val!= q->val)
+        return x;
+    Node* y = commonAncestorHelp(root->right, p, q);
+    if (y != nullptr && y->val != p->val && y->val!= q->val)
+        return y;
+    
+    if (x != nullptr && y!=nullptr)
```
Comment (Brusi) [inline assignment2.cpp:69]: Possible comment:
`// Found target nodes in both children subtrees. root is the lowest common ancestor.`

File: assignment2.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    return cover(root->left, target) || cover(root->right, target);
+}
+
+Node* commonAncestorHelp(Node* root, Node* p, Node* q)
+{
+    if (root ->val== p->val && root->val == q->val) return root;
+    if (root==nullptr) return nullptr;
+    
+    Node* x = commonAncestorHelp(root->left, p, q);
+    if (x != nullptr && x->val != p->val && x ->val!= q->val)
```
Comment (Brusi) [inline assignment2.cpp:63]: What is this check for?
It checks that `x` is neither `p` or `q`, meaning, this is not one of the required nodes, so it must be that we found the lowest common ancestor, so no need to continue searching.
But what if the lowest common ancestor is one of the target nodes itself? Then this check will fail.
For example in the tree:
```
1
 \
  2
 /
3
```
When searching for nodes `2` and `3`, this check will return false when root is node 2 (and `x` is node 3).

This does not seem to make your algorithm incorrect (I could not find an example where your code returns a wrong result). However, each line of code needs a reason.
If you did this for optimization, so it would not search for more nodes after it had found the common ancestor - then  this reason is good enough. But it is still very confusing that this optimization works only in some of the times.
So I suggest that you either find a different way to check if you already found the root (so you will always be correct and apply this optimization), or remove it to make your code shorter or clearer.
Consider adding comments to make wierd places clear.
Also, you can extract unclear lines to functions with names that indicate what the code does.

File: assignment2.cpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+bool cover(Node* root, Node* target)
+{
+    if (root == nullptr)
+        return false;
+    if (root->val == target->val)
+        return true;
+    return cover(root->left, target) || cover(root->right, target);
+}
+
+Node* commonAncestorHelp(Node* root, Node* p, Node* q)
```
Comment (Brusi) [inline assignment2.cpp:57]: Since this function is not straightforward (it combines searching for the children AND returning the root node), I recommend adding comments that clarifies what it does, and note on specific code lines what it does.
For example, you can document this function:
`// Recursively searches for nodes p and q. On the way back, if both children contain q or p, the root node will be returned all the way up.`

(This was an example, you can change the wording or the entire comment)

**Review by Brusi — COMMENTED** [review]
Well done, see comments.
You can add the rest of the team as collaborators so they can peer-review your code:
elenial
matinehAkhlaghinia
maryb14
katabana
Mais051

File: assignment3.cpp [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+class Dictionary{
+public:
+    bool isWord(string s);
+    bool isPrefix(string s);
+};
+
+void wordSearchHelper(int i, int j, int n, int m, char** grid, bool** used,  Dictionary* dict, string currStr, set<string>* wordSet)
+{
+    if(used[i][j]==false)
```
Comment (Brusi) [inline assignment3.cpp:23]: Tip:
You can return early to avoid nesting:
```C++
if (!used[i][j]) {  // Prefer this form over comparing booleans to true/false.
   return;
}
currStr += grid[i][j];
if (!dict->isPrefix(currStr)) {
  return;
}
```
Instead of making all the function body inside a big "if".

There is one downside in returning early - then it is possible to forget to set `used[i][j]` back to false. If you choose to return early, make sure that `used` is in the correct state when returning.

File: assignment3.cpp [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        used[i][j]=true;
+        currStr+=grid[i][j];
+        if(dict->isPrefix(currStr))
+        {
+            if(dict->isWord(currStr))
+                wordSet->insert(currStr);
+            if(i>0)
```
Comment (Brusi) [inline assignment3.cpp:31]: Tip for making shorter code:
Instead of comparing `i` and `j` before each call, you can move these checks to the start of the function, and return early in case `i` or `j` exceed the limits.

```C++
void wordSearchHelper(int i, int j, ...) {
    if (i < 0 || i >= n || j < 0 || j >= m) {
        return;
    }
    if (!used[i][j]) { 
    ......
```

Then you can just write all the recursive calls one after the other:
```C++
  wordSearchHelper(i-1,j-1, n, m, grid,used, dict,currStr,wordSet);
  wordSearchHelper(i-1,j, n, m, grid,used, dict,currStr,wordSet);
  wordSearchHelper(i-1,j+1, n, m, grid,used, dict,currStr,wordSet);
  ...
```

Or use some kind of a loop to do it.

File: assignment3.cpp [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        }
+        used[i][j]=false;
+    }
+}
+
+set<string>* wordSarch(int n, int m, char** grid, Dictionary* dict)
+{
+    bool used[n][m];
+    fill(used[0], used[0] + n * m, 0);
+    set<string>* wordSet=new set<string>();
```
Comment (Brusi) [inline assignment3.cpp:60]: Tip to avoid dealing with memory issues:
No need to allocate it dynamically. It is OK to create it on the stack and return it by value.
So you can do here: `set<string> wordsSet;`
And later Just make sure you send it to the helper by address:
`wordsSearchHelper(... , &wordSet);`

File: assignment3.cpp [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    string currStr="";
+    for(int i=0; i<n; i++)
+        for(int j=0; j<m; j++)
+            wordSearchHelper(i,j, n, m, grid, (bool**)used, dict, currStr, wordSet);
```
Comment (Brusi) [inline assignment3.cpp:64]: Passing 'used' this way may compile, but it will not work -- a 2D array is substantially different than a pointer to a pointer.
Since passing a 2D array to a function in C/C++ is impossible without knowing the array size in compile time (which you don't in this assignment), I suggest 2 alternatives:
1. Create `used` as an array of arrays. Allocate a basic array of size `n`, and then fill each slot with an array of size `m`. Then you can pass it to the function as is and it would work. Just remember to deallocate it all when you are done using it :)
2. Use a vector of vectors, which is a lot more comfortable in my opinion:
`vector<vector<bool>> used(n, vector<bool>(m, false));`
The first parameter in a vector constructor is its size, and the second is the default value. So the command above initializes a vector of size `n`, where each cell contains a vector of size `m` initialized to `false`.

Same applies to the `grid`. But I did not see its testing code, so if you pass it as an array of arrays then it is good as is. Though I still recommend vectors for it ;)

File: assignment4.cpp [hunk 12]
```diff
@@ -0,0 +1,79 @@
+//
+//  assignment4.cpp
+//  assignment4
+//
+//  Created by Tamara Sekularac on 6/25/17.
+//  Copyright © 2017 Tamara Sekularac. All rights reserved.
+//
+
+
```
Comment (Brusi) [inline assignment4.cpp:9]: OK, after a lot of reading, I finally understood your algorithm!
It is very nice!
It greedily creates islands by connecting adjacent map cells, and assigning a running "island index" to any cell.
When a cell can connect to one adjacent cell, it gets its "island index".
When a cell can connect to 2 different islands (top and left cells are different islands), then you have some sort of an "islands tree" where an island index points to another island to indicate that they are actually connected. It always sets the island with the larger index as parent (to avoid circles), and eventually it counts connected island indices as a single island.

Very good solution for this assignment. :+1: 

But as I said before, the code was very hard for me to understand. I believe that you can improve it a lot by using functions, giving more indicative variable names, avoiding code duplications and nesting whenever you can, and adding comments.
(For example, a sentence or two that explains your algorithm, as I wrote here above, would save me time in trying to understand this code. Of course it is not about my time, but about how your code is the most readable and maintainable by everyone who reads it :) )

I'll elaborate more in the following comments.

File: assignment4.cpp [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+int numberOfIslands(int n, int m, bool** map)
+{
+    int mapNum[n][m];
+    int num=0;
```
Comment (Brusi) [inline assignment4.cpp:13]: Instead of name "num", which does not say a lot about what this number actually holds, what about calling it "num_islands" or "island_index"?

File: assignment4.cpp [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+//  assignment4
+//
+//  Created by Tamara Sekularac on 6/25/17.
+//  Copyright © 2017 Tamara Sekularac. All rights reserved.
+//
+
+
+int numberOfIslands(int n, int m, bool** map)
+{
+    int mapNum[n][m];
```
Comment (Brusi) [inline assignment4.cpp:12]: Maybe call it "cell_to_island_index_map" ?
Or something shorter like "index_map", but then add a comment that says what it holds.
e.g. `// For each cell in the map, holds the island index of this cell`.

File: assignment4.cpp [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+    int num=0;
+    int islandConnection[n*m];
```
Comment (Brusi) [inline assignment4.cpp:14]: This is a good name :+1: 

I see that you set its size to `n*m` which is a top limit of possible size it can get. This is ok.
Another solution is to use a `std::vector`, and increase its size every time you start a new island. It would 'waste' less space, but mostly it will be clear that all of its cells have content (and no cells are 'empty' or contain garbage).

File: assignment4.cpp [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    if(map[0][0])
```
Comment (Brusi) [inline assignment4.cpp:15]: OK, starting :)

You handle the top left cell first; then the top row; then the left column, then the rest.
This is basically ok, but creates a lot of duplication. e.g. `num++; map[i][j]=num` repeates a lot.
One thing that can save most of it, is making a function to accessing `map` which is index-safe.
```C++
bool mapHasLand(bool** map, int i, int j) {
    return i >= 0 && j >= 0 && map[i][j];
}
```
This way you can handle every cell the same way, and the top row and left column will need no special treatment:
```C++
for (int i=0; i < n; i++) {
  for (int j=0; j < m; j++) {
    if (mapHasLand(map, i - 1, j)) {
    .....
  }
}
```

File: assignment4.cpp [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+        {
+            if(map[i-1][0])
+            mapNum[i][0]=mapNum[i-1][i];
+            else
+            {
+                num++;
+                mapNum[i][0]=num;
+            }
+        }
+        for(int j=1; j<m; j++)
```
Comment (Brusi) [inline assignment4.cpp:44]: I think you have a bug here.
Shouldn't `i` start with 1?
It was best to first finish the loop of the first column, and then start the loop for the "inner cells":
```C++
for(int i=0; i<n; i++) {  // The loop from line 32
   ...
}  // finish it
for(int i=1; i<n; i++) {  // Then start the inner loop
  for(int j=1; j<m; j++) {
   ...
}

File: assignment4.cpp [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+            if(map[i][j])
```
Comment (Brusi) [inline assignment4.cpp:45]: OK, here the actual main loop starts.
Remember 'return early' from assignment 3? So you can apply this here using `continue`. e.g.
```C++
for(int i=1; i<n; i++) {
  for(int j=1; j<m; j++) {
    if (!map[i][j]) {
      continue;
    }
    if (map[i-1][j]) ...
  }
}

File: assignment4.cpp [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+            {
+                if(map[i-1][j])
```
Comment (Brusi) [inline assignment4.cpp:47]: OK, this is the heart of the algorithm.
Now you have several "cases" you can handle. Let's make the code indicate those cases:
```C++
for(int i=1; i<n; i++) {
  for(int j=1; j<m; j++) {
    if (!map[i][j]) continue; // This cell has no land.
    if (!map[i-1][j] && !map[i][-1j]) {  // This cell starts a new island
        mapNum[i][j] = ++num;
        continue;
    }
    if (map[i-1][j] && !map[i][-1j]) {  // Only cell to the top has land
        mapNum[i][j] = mapNum[i-1][j];
        continue;
    }
    if (!map[i-1][j] && map[i][j-1]) {  // Only cell to the left an island
        mapNum[i][j] = mapNum[i][j-1];
        continue;
    }
    // Now we know that we have land in both cells to the top and left. Handle the rest of the cases...
    ....
  }
}
```
This way, each different case and what you do in it, is much clearer. Take the advice from above to check cells safely with a function, and you can check each case in only one place in the code; and this is the recipe for a good and readable code!

Note that there is almost no 'else' in this example, and no 'if' inside another 'if'. There is nothing bad with it in general, but overusing them may make the code hard to follow, while making the code more 'flat' is easier to follow.

File: assignment4.cpp [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+                    else
+                    {
+                        num++;
+                        mapNum[i][j]=num;
+                    }
+                }
+    }
+    int rez=0;
+    for(int i=1; i<=num; i++)
+        if(i==islandConnection[i])
```
Comment (Brusi) [inline assignment4.cpp:75]: I think you did not set a default value to `islandConnection` in the case where you don't specifically add a pointer.
You should make an island index point at itself each time you increase `num`:
```C++
num++;
islandConnection[num] = num;
mapNum[i][j] = num;
```
Or better, if `islandConnection` is a vector:
`islandConnection.push_back(num);`

**Review by Brusi — COMMENTED** [review]
Solutions are correct and very good for the assignments. Well done!

My two main remarks:
1. I had a few comments about how to improve readability of the code and make it clearer for the reader (especially for assignment 4). I hope I gave you concrete points about what can you improve the code.
2. I found a few bugs in the code, that I believe would be very easy to find when testing the code. It is very much recommended that you test your code on several examples, including some interesting corner cases. You can learn a lot from it even without anyone reviewing your code!

File: assignment6.cpp [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+    int nextOne=1, posTo, posFrom, carNum;
+    int curP[maxN], endP[maxN];
+    for(int i=0; i<N;i++)
+    {
+        curP[startPos[i]]=i;
+        endP[endPos[i]]=i;
+    }
+    while(curP[0]!=endP[0])
+    {
+        posTo=curP[0];
```
Comment (katabana) [inline assignment6.cpp:33]: I think variable names could be better. You might know what it means only when you know the whole assignment.

File: assignment6.cpp [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+    {
+        if(curP[nextOne]!=endP[nextOne])
+        {
+            posTo=curP[0];
+            carNum=nextOne;
+            posFrom=curP[nextOne];
+            curP[0]=posFrom;
+            curP[carNum]=posTo;
+            cout<<"move from "<<posFrom<<" to "<<posTo<<endl;
+            while(curP[0]!=endP[0])
```
Comment (katabana) [inline assignment6.cpp:50]: You used exactly the same code in another while loop. You could make it function ;) And again - `while`. Although its outcome is the same, I think it might confuse someone who reads it. At least I was a little bit confused when I saw this :)

File: assignment6.cpp (hunk 21 above)
Comment (katabana) [reply assignment6.cpp:33]: If I understand it well, in this loop you make '0' stay on its final position? However, I do not understand why you use `while` instead of simple `if`.

File: assignment6.cpp [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+//
+//  Created by Tamara Sekularac on 7/21/17.
+//  Copyright © 2017 Tamara Sekularac. All rights reserved.
+//
+
+#include <list>
+#include <set>
+#include <unordered_map>
+#include <string>
+#include <queue>
```
Comment (katabana) [inline assignment6.cpp:13]: I do not think you use so many structures in this program, therefore those includes are redundant ;)

**Review by katabana — COMMENTED** [review]
In general your code is easy and I think it is correct. I agree with said complexity :)
